pygmo/_py_algorithms.py

```diff
@@ -0,0 +1,562 @@
+
+
+class scipy:
+    """
+    This class is a user defined algorithm (UDA) providing a wrapper around the function scipy.optimize.minimize.
+
+    This wraps several well-known local optimization algorithms:
+     - Nelder-Mead
+     - Powell
+     - CG
+     - BFGS
+     - Newton-CG
+     - L-BFGS-B
+     - TNC
+     - COBYLA
+     - SLSQP
+     - trust-constr
+     - dogleg
+     - trust-ncg
+     - trust-exact
+     - trust-krylov
+
+    These methods are mostly variants of gradient descent. Some of them require a gradient and will throw
+    an error if invoked on a problem that does not offer one.
+    Constraints are only supported by methods COBYLA, SLSQP and trust-constr.
+
+    Example:
+
+    >>> import pygmo as pg
+    >>> prob = pg.problem(pg.rosenbrock(10))
+    >>> pop = pg.population(prob=prob, size=1, seed=0)
+    >>> pop.champion_f[0]
+    929975.7994682974
+    >>> scp = pg.algorithm(pg.scipy(method="L-BFGS-B"))
+    >>> result = scp.evolve(pop).champion_f
+    >>> result[0]
+    1.13770...
+    >>> pop.problem.get_fevals()
+    55
+    >>> pop.problem.get_gevals()
+    54
+    """
+
+    def _maybe_jit(func):
+        """
+        This function tries to import the just-in-time compiler from numba and apply it to the passed function.
+        If the import fails, the argument is returned unchanged.
+        """
+        try:
+            from numba import jit
+
+            return jit(nopython=True)(func)
+        except ModuleNotFoundError:
+            return func
+
+    @staticmethod
+    def _generate_gradient_sparsity_wrapper(
+        func, idx: int, dim: int, sparsity_func, invert_sign=False
+    ):
+        """
+        A function to extract a sparse gradient from a pygmo problem to a dense gradient expectecd by scipy.
+
+        Pygmo convention is to include problem constraints into its fitness function. The same applies to the gradient.
+        The scipy.optimize.minimize function expects a separate callable for each constraint, this function creates a wrapper that extracts a requested dimension.
+        It also transforms the sparse gradient into a dense representation.
+
+        Args:
+
+            func: the gradient callable
+            idx: the requested dimension.
+            dim: number of arguments for callable func and the resulting gradient
+            sparsity_func: a callable giving the sparsity pattern. Use problem.gradient_sparsity.
+            invert_sign: whether all values of the gradient should be multiplied with -1. This is necessary for inequality constraints, where the feasible side is interpreted the opposite way by scipy and pygmo.
+
+        Returns:
+
+            a callable that passes all arguments to the gradient callable func and returns the dense gradient at dimension idx
+
+        Raises:
+
+            unspecified: any exception thrown by sparsity_func
+
+
+        """
+        sparsity_pattern = sparsity_func()
+        # TODO: check whether pattern, dim and idx fit together
```

Comment:
Is this TODO still relevant?

Reply:
Not really. Since the function is only called internally, we don't have to be as vigilant about the validity of the passed arguments.